src/connection_fallback.ts

```diff
@@ -0,0 +1,205 @@
+  cancelToken?: CancelTokenSource;
+
+  constructor({
+    client,
+  }: {
+    client: StreamChat<AttachmentType, ChannelType, CommandType, EventType, MessageType, ReactionType, UserType>;
+  }) {
+    this.client = client;
+    this.state = ConnectionState.Init;
+    this.consecutiveFailures = 0;
+
+    addConnectionEventListeners(this._onlineStatusChanged);
+  }
+
+  _log(msg: string, extra: UR = {}, level: LogLevel = 'info') {
+    this.client.logger(level, 'WSConnectionFallback:' + msg, { tags: ['connection_fallback', 'connection'], ...extra });
+  }
+
+  _setState(state: ConnectionState) {
+    this._log(`_setState() - ${state}`);
+
+    // transition from connecting => connected
+    if (this.state === ConnectionState.Connecting && state === ConnectionState.Connected) {
+      //@ts-expect-error
+      this.client.dispatchEvent({ type: 'connection.changed', online: true });
+    }
+
+    if (state === ConnectionState.Closed || state === ConnectionState.Disconnected) {
+      //@ts-expect-error
+      this.client.dispatchEvent({ type: 'connection.changed', online: false });
+    }
+
+    this.state = state;
+  }
+
+  /** @private */
+  _onlineStatusChanged = (event: { type: string }) => {
+    this._log(`_onlineStatusChanged() - ${event.type}`);
+
+    if (event.type === 'offline') {
+      this._setState(ConnectionState.Closed);
+      this.cancelToken?.cancel('disconnect() is called');
+      this.cancelToken = undefined;
+      return;
+    }
+
+    if (event.type === 'online' && this.state === ConnectionState.Closed) {
+      this.connect(true);
+    }
+  };
+
+  /** @private */
+  _req = async <T = UR>(params: UnknownType, config: AxiosRequestConfig, retry: boolean): Promise<T> => {
+    if (!this.cancelToken && !params.close) {
+      this.cancelToken = axios.CancelToken.source();
+    }
+
+    try {
+      const res = await this.client.doAxiosRequest<T>(
+        'get',
+        (this.client.baseURL as string).replace(':3030', ':8900') + '/longpoll', // replace port if present for testing with local API
+        undefined,
+        {
+          config: { ...config, cancelToken: this.cancelToken?.token },
+          params,
+        },
+      );
+
+      this.consecutiveFailures = 0; // always reset in case of no error
+      return res;
+    } catch (err) {
+      this.consecutiveFailures += 1;
+
+      if (retry && isErrorRetryable(err)) {
+        this._log(`_req() - Retryable error, retrying request`);
+        await sleep(retryInterval(this.consecutiveFailures));
+        return this._req<T>(params, config, retry);
+      }
+
+      throw err;
+    }
+  };
+
+  /** @private */
+  _poll = async () => {
+    while (this.state === ConnectionState.Connected) {
+      try {
+        const data = await this._req<{
+          events: Event<AttachmentType, ChannelType, CommandType, EventType, MessageType, ReactionType, UserType>[];
+        }>({}, { timeout: 30000 }, true); // 30s => API responds in 20s if there is no event
```

Comment:
27s to be exact

Reply:
the longpoll responds with empty event list after 20s of no event

Reply:
true, `events := client.ConsumeEvents(10, 500*time.Millisecond, 20*time.Second)`

I remembered 27 for some reason